### ISSUE TITLE
Exception is thrown with latest dev sdk: type 'Future<dynamic>' is not a subtype of type 'FutureOr  ...' 

### ISSUE THREAD

**Issue by supermuka** [dart-lang/rpc#121]
After upgrading Dart to 2.0.0-dev.57.0, the following exception is been thrown on the browser console:

> `EXCEPTION: Type 'List<dynamic>' should be 'List<UserProfileOrganization>' to implement expected type 'FutureOr<List<UserProfileOrganization>>'`

It is being use `rpc package` (last version) to implement REST APIs and this exception occurs into client-side stub file generated by _**rpc tool**_.

Note: `UserProfileOrganization` is a _message class_ (my code) used to implement the APIs with rpc.

To solve this problem, the use of _Dart2js_ was suggested instead _DCC_, as discussed in Gitter on [mai 23 15:28](https://gitter.im/dart-lang/server?at=5b05b2d5aaafa25932ae3fbd).

Since then, I'm using _Dart2js_ including -r parameter on build command line as _pub run build_runner serve -r_. This way the above exception is no longer thrown.

But it will be interesting if _rpc_ works also with DDC.

**Comment by enyo** [comment]
I just ran into this problem. Is there a solution yet? Using `dart2js` is not really a solution for development.

**Comment by enyo** [comment]
I don't quite get the problem either... The generated code in question (if I'm not mistaken) is this:

```dart
oAuthAccounts = _json["oAuthAccounts"].map((value) => new OAuthAccount.fromJson(value)).toList();
```

where `oAuthAccounts` is of type `core.List<OAuthAccount>`. Why is that not working?

**Comment by mkustermann** [comment]
So this is a Dart 2.0 runtime type error: a `List<dynamic>` is not a subtype of `List<OAuthAccount>`.

We have updated `package:discoveryapis_generator` to produce valid Dart 2.0 code. Could you verify that after a `pub upgrade`, your `.packages` file contains 
```
discoveryapis_generator:file:///.../.pub-cache/hosted/pub.dartlang.org/discoveryapis_generator-0.9.7/lib/
_discoveryapis_commons:file:///.../.pub-cache/hosted/pub.dartlang.org/_discoveryapis_commons-0.1.6/lib/
```

If not, there might be some 3rd party dependency (or your Dart SDK) requiring a lower version of `package:discoveryapis_generator`, which basically means you're using a generator which is not Dart 2.0 compliant.

Currently the code should be generated e.g. like [this example from package:googleapis](https://github.com/dart-lang/googleapis/blob/master/generated/googleapis/lib/adexchangebuyer/v1_3.dart#L2325):
```dart
  DirectDealsList.fromJson(core.Map _json) {
    if (_json.containsKey("directDeals")) {
      directDeals = (_json["directDeals"] as core.List)
          .map<DirectDeal>((value) => new DirectDeal.fromJson(value))
          .toList();
```
By using `.map<DirectDeal>` it should get the right runtime type.

**Comment by enyo** [comment]
Thanks @mkustermann ! You're right. The reason we got the old version, is that I forgot that `pub run rpc:generate` uses the globally activated `discoveryapis_generator` so I had to run `pub global activate discoveryapis_generator` to get the latest version.

Is there a way to get `rpc:generate` to run with the local package dependency instead of the globally activated one? This would prevent such a problem to reoccur in the future...

Other than that, I think that this issue can be closed.

**Comment by mkustermann** [comment]
This is really a question for the pub team, but so far I had the impression that:

* You use a local version via
```
$ cat pubspec.yaml
dependencies:
  rpc: any
$ pub upgrade
$ pub run rpc:generate
```

* You use a global version via
```
$ pub global activate rpc
$ pub global run rpc:generate
```

**Comment by enyo** [comment]
Yes, but `pub run rpc:generate` seems to only use the globally activated `discoveryapis_generator` and not the locally installed one. That's why I thought it might be `rpc` specific.

**Comment by mkustermann** [comment]
> Yes, but pub run rpc:generate seems to only use the globally activated discoveryapis_generator and not 
> the locally installed one. That's why I thought it might be rpc specific.

Aha. So you're saying `pub run rpc:generate` does not use the `discoveryapis_generator` specified in the `.packages` file but rather the one which is globally activated?

If that's the case, could you please file a bug with pub? (cc @nex3 )

**Comment by supermuka** [comment]
An addendum on the errors considering _SDK 2.0.0-dev.64.1_  and _rpc package_ 0.5.10 (today are the last version).

When I try to generate the client side APIs, a new error is occurring:

> `pub run rpc:generate client -i lib/server/augeapi.dart -o lib/client -p 8091`

> Unhandled exception:
> type 'Future<dynamic>' is not a subtype of type 'FutureOr<List<DescriptionImportPair>>'
> #0      ClientApiGenerator.generateDiscoveryWithImports (file:///D:/Users/Samuel/AppData/Roaming/Pub/Cache/hosted/pub.dartlang.org/rpc-0.5.10/bin/generate.dart:204:12)
> #1      _AsyncAwaitCompleter.start (dart:async/runtime/libasync_patch.dart:49:6)
> #2      ClientApiGenerator.generateDiscoveryWithImports (file:///D:/Users/Samuel/AppData/Roaming/Pub/Cache/hosted/pub.dartlang.org/rpc-0.5.10/bin/generate.dart:203:67)
> #3      main (file:///D:/Users/Samuel/AppData/Roaming/Pub/Cache/hosted/pub.dartlang.org/rpc-0.5.10/bin/generate.dart:122:47)
> #4      _AsyncAwaitCompleter.start (dart:async/runtime/libasync_patch.dart:49:6)
> #5      main (file:///D:/Users/Samuel/AppData/Roaming/Pub/Cache/hosted/pub.dartlang.org/rpc-0.5.10/bin/generate.dart:85:5)
> #6      _startIsolate.<anonymous closure> (dart:isolate/runtime/libisolate_patch.dart:277:32)
> #7      _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:165:12)
>

**Comment by enyo** [comment]
@supermuka for me it's working fine now. I'm on `dev.63` though because `dev.64.1` had problems with globally installed packages. Try reverting to `dev.63` and see if that changes anything.

**Comment by supermuka** [comment]
@enyo, for me works on `dev.63` too. But... 

I think this package deserved more attention in its maintenance, focusing on new versions and suggestions for improvement. I say this because we can see many issues open, which worries me, because _this package is one of the most useful and nice things_ that the _Dart Team_ has already done to facilitate the implementation of RESTfull APIs. I would love to see this package taking special care in its evolution.

**Comment by supermuka** [comment]
@mkustermann, on version SDK `dev.65` and `rpc package 0.5.10` when `pub run rpc:generate` is ran, the following error continue being thrown:

> `type 'Future<dynamic>' is not a subtype of type 'FutureOr<List<DescriptionImportPair>>'`

If there is a suggested workaround, I'd be grateful!

Note: because of other package dependencies, it is necessary to use the latest version of the SDK.

**Comment by supermuka** [comment]
Changed the title for this issue, considering that last version of the SDK thrown exception to Dart2js too.

**Comment by mkustermann** [comment]
I've now looked at this and it appears that, after fixing some Dart 2.0 issues, it is blocked on the fact that `Isolate.spawnUri` has a bug in the Dart 2.0 version, where it doesn't support spawning from http urls anymore. 

So this is blocked on https://github.com/dart-lang/sdk/issues/33741

**Comment by jcollins-g** [comment]
Fixed in #124.
